### PR TITLE
feat(compiler-cli): add option to ng-xi18n to specify the source lang…

### LINKED ISSUE #12303: Extend the ng-xi18n tool with a sourceLanguage argument and a targetLanguages argument
**I'm submitting a ...**

```
[x] feature request
```

**Current behavior**
Running the `ng-xi18n` tool will create a `messages.xlf` file with `source-language="en"` attribute and no `target-language="..."` attribute (https://en.wikipedia.org/wiki/XLIFF).

**Expected behavior**
Please extend the ng-xi18n tool with a `sourceLanguage` argument and a `targetLanguages` argument.
The `sourceLanguage` argument will control the `source-language=` attribute.
For every language provided by `targetLanguages` a `messages.XX.xlf` will be created and for each created file the `target-language="..."` attribute will be set.

**What is the motivation / use case for changing the behavior?**
Create multiple `messages.XX.xlf` files with correct `source-language=` attribute and correct `target-language="..."` attribute.

**Please tell us about your environment:**
Linux
- **Angular version:** 2.1.0
- **Browser:** All
- **Language:** TypeScript 2.0.3
- **Node (for AoT issues):** 6.7.0

### PR DESCRIPTION
**Pull request by marclaval**
…uage

Fixes #12303

@vicb I tried to guess the correct format for XMB.
Can you please review this?

### CODE REVIEW

File: modules/@angular/compiler-cli/integrationtest/test/test_ngtools_api.ts [hunk 1]
```diff
@@ -131,6 +131,7 @@ function i18nTest() {
         compilerOptions: config.parsed.options, program, host,
         angularCompilerOptions: config.ngOptions,
         i18nFormat: 'xlf',
+        i18nSourceLanguage: null,
```
Comment (vicb) [inline modules/@angular/compiler-cli/integrationtest/test/test_ngtools_api.ts]: locale ?

File: modules/@angular/compiler/src/i18n/serializers/xliff.ts [hunk 2]
```diff
@@ -59,7 +59,11 @@ export class Xliff extends Serializer {
 
     const body = new xml.Tag('body', {}, [...transUnits, new xml.CR(4)]);
     const file = new xml.Tag(
-        'file', {'source-language': _SOURCE_LANG, datatype: 'plaintext', original: 'ng2.template'},
+        'file', {
+          'source-language': sourceLanguage || _SOURCE_LANG,
+          datatype: 'plaintext',
+          original: 'ng2.template'
```
Comment (vicb) [inline modules/@angular/compiler/src/i18n/serializers/xliff.ts]: nit: trailing coma

File: modules/@angular/compiler/src/i18n/serializers/xliff.ts [hunk 3]
```diff
@@ -59,7 +59,11 @@ export class Xliff extends Serializer {
 
     const body = new xml.Tag('body', {}, [...transUnits, new xml.CR(4)]);
     const file = new xml.Tag(
-        'file', {'source-language': _SOURCE_LANG, datatype: 'plaintext', original: 'ng2.template'},
+        'file', {
+          'source-language': sourceLanguage || _SOURCE_LANG,
```
Comment (vicb) [inline modules/@angular/compiler/src/i18n/serializers/xliff.ts]: rename to _DEFAULT_SOURCE_LANG

File: modules/@angular/compiler/test/i18n/serializers/xmb_spec.ts [hunk 4]
```diff
@@ -45,7 +45,7 @@ export function main(): void {
 
 <!ELEMENT ex (#PCDATA)>
 ]>
-<messagebundle>
+<messagebundle lang="en">
```
Comment (vicb) [inline modules/@angular/compiler/test/i18n/serializers/xmb_spec.ts]: as you can see from the DTD above there is no lang on message bundles - it is assumed to be en.
Translations bundles have a lang attribute.

**Review by vicb — DISMISSED** [review]
- I think locale might be better than sourceLanguage ?
- xmb does not have a lang,
- few other nits

File: modules/@angular/compiler/src/i18n/serializers/xliff.ts [hunk 5]
```diff
@@ -29,7 +29,7 @@ const _UNIT_TAG = 'trans-unit';
 // http://docs.oasis-open.org/xliff/v1.2/os/xliff-core.html
 // http://docs.oasis-open.org/xliff/v1.2/xliff-profile-html/xliff-profile-html-1.2.html
 export class Xliff extends Serializer {
-  write(messages: i18n.Message[]): string {
+  write(messages: i18n.Message[], locale: string): string {
```
Comment (vicb) [inline modules/@angular/compiler/src/i18n/serializers/xliff.ts]: locale?:string = _DEFAULT_SOURCE_LANG

File: modules/@angular/compiler/src/i18n/serializers/xliff.ts [hunk 6]
```diff
@@ -59,7 +59,11 @@ export class Xliff extends Serializer {
 
     const body = new xml.Tag('body', {}, [...transUnits, new xml.CR(4)]);
     const file = new xml.Tag(
-        'file', {'source-language': _SOURCE_LANG, datatype: 'plaintext', original: 'ng2.template'},
+        'file', {
+          'source-language': locale || _DEFAULT_SOURCE_LANG,
```
Comment (vicb) [inline modules/@angular/compiler/src/i18n/serializers/xliff.ts]: and remove here

File: modules/@angular/compiler/src/i18n/serializers/xmb.ts [hunk 7]
```diff
@@ -38,7 +38,7 @@ const _DOCTYPE = `<!ELEMENT messagebundle (msg)*>
 <!ELEMENT ex (#PCDATA)>`;
 
 export class Xmb extends Serializer {
-  write(messages: i18n.Message[]): string {
+  write(messages: i18n.Message[], locale: string): string {
```
Comment (vicb) [inline modules/@angular/compiler/src/i18n/serializers/xmb.ts]: locale?:string

File: modules/@angular/compiler-cli/src/extractor.ts [hunk 8]
```diff
@@ -27,14 +27,14 @@ export class Extractor {
       public host: ts.CompilerHost, private ngCompilerHost: CompilerHost,
       private program: ts.Program) {}
 
-  extract(formatName: string): Promise<void> {
+  extract(formatName: string, locale: string): Promise<void> {
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/extractor.ts]: ?:string

File: modules/@angular/compiler-cli/src/ngtools_api.ts [hunk 9]
```diff
@@ -59,6 +59,7 @@ export interface NgTools_InternalApi_NG2_ExtractI18n_Options {
   host: ts.CompilerHost;
   angularCompilerOptions: AngularCompilerOptions;
   i18nFormat: string;
+  locale: string;
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/ngtools_api.ts]: optional

**Comment by ocombe** [comment]
@vicb I made the changes that you requested, it should be good to merge now

**Comment by ocombe** [comment]
The failed tests have nothing to do with the PR by the way, just merging master into the branch and it runs smoothly on my machine

File: modules/@angular/compiler-cli/src/extractor.ts [hunk 10]
```diff
@@ -27,14 +27,14 @@ export class Extractor {
...
-  extract(formatName: string): Promise<void> {
+  extract(formatName: string, locale?: string): Promise<void> {
     // Checks the format and returns the extension
     const ext = this.getExtension(formatName);
 
     const promiseBundle = this.extractBundle();
 
     return promiseBundle.then(bundle => {
-      const content = this.serialize(bundle, ext);
+      const content = this.serialize(bundle, ext, locale);
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/extractor.ts]: `locale || null`

File: modules/@angular/compiler-cli/src/extractor.ts [hunk 11]
```diff
@@ -59,7 +59,7 @@ export class Extractor {
         serializer = new compiler.Xliff();
     }
 
-    return bundle.write(serializer);
+    return bundle.write(serializer, locale);
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/extractor.ts]: locale || null

File: modules/@angular/compiler/src/i18n/message_bundle.ts [hunk 12]
```diff
@@ -47,7 +47,7 @@ export class MessageBundle {
   // The public (serialized) format might be different, see the `write` method.
   getMessages(): i18n.Message[] { return this._messages; }
 
-  write(serializer: Serializer): string {
+  write(serializer: Serializer, locale?: string): string {
```
Comment (vicb) [inline modules/@angular/compiler/src/i18n/message_bundle.ts]: locale: string | null

File: modules/@angular/compiler/src/i18n/serializers/serializer.ts [hunk 13]
```diff
@@ -12,7 +12,7 @@ export abstract class Serializer {
   // - The `placeholders` and `placeholderToMessage` properties are irrelevant in the input messages
   // - The `id` contains the message id that the serializer is expected to use
   // - Placeholder names are already map to public names using the provided mapper
-  abstract write(messages: i18n.Message[]): string;
+  abstract write(messages: i18n.Message[], locale: string): string;
```
Comment (vicb) [inline modules/@angular/compiler/src/i18n/serializers/serializer.ts]: string | null

File: modules/@angular/compiler/src/i18n/serializers/xliff.ts [hunk 14]
```diff
@@ -29,7 +29,7 @@ const _UNIT_TAG = 'trans-unit';
 // http://docs.oasis-open.org/xliff/v1.2/os/xliff-core.html
 // http://docs.oasis-open.org/xliff/v1.2/xliff-profile-html/xliff-profile-html-1.2.html
 export class Xliff extends Serializer {
-  write(messages: i18n.Message[]): string {
+  write(messages: i18n.Message[], locale?: string): string {
```
Comment (vicb) [inline modules/@angular/compiler/src/i18n/serializers/xliff.ts]: locale: string|null

File: modules/@angular/compiler/src/i18n/serializers/xmb.ts [hunk 15]
```diff
@@ -38,7 +38,7 @@ const _DOCTYPE = `<!ELEMENT messagebundle (msg)*>
 <!ELEMENT ex (#PCDATA)>`;
 
 export class Xmb extends Serializer {
-  write(messages: i18n.Message[]): string {
+  write(messages: i18n.Message[], locale: string = ''): string {
```
Comment (vicb) [inline modules/@angular/compiler/src/i18n/serializers/xmb.ts]: locale: string|null

File: modules/@angular/compiler/src/i18n/serializers/xtb.ts [hunk 16]
```diff
@@ -19,7 +19,9 @@ const _TRANSLATION_TAG = 'translation';
 const _PLACEHOLDER_TAG = 'ph';
 
 export class Xtb extends Serializer {
-  write(messages: i18n.Message[]): string { throw new Error('Unsupported'); }
+  write(messages: i18n.Message[], sourceLanguage: string): string {
```
Comment (vicb) [inline modules/@angular/compiler/src/i18n/serializers/xtb.ts]: |null

File: modules/@angular/compiler-cli/src/extractor.ts [hunk 17]
```diff
@@ -47,7 +47,7 @@ export class Extractor {
     return this.ngExtractor.extract(files);
   }
 
-  serialize(bundle: compiler.MessageBundle, ext: string): string {
+  serialize(bundle: compiler.MessageBundle, ext: string, locale?: string|null): string {
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/extractor.ts]: should be `locale:` ie not optional

**Review by vicb — APPROVED** [review]

**Pull request closed without merging**